### ISSUE TITLE
Workflow hangs indefinitely during "ContainerCreating" if the Pod or Node unexpectedly dies

### ISSUE THREAD

**Issue by rysunder** [argoproj/argo-workflows#5543]
## Summary 
When a node is drained or pods are evicted from a node during the "ContainerCreating" step, the argo-workflow goes into a zombie state; perpetually waiting for the non-existent container to create.

Ideally at this stage argo would recognize the pod/node is no longer present and retry the task or report a failure for that workflow.  

The easiest way to reproduce this error is to delete the pod during the "ContainerCreating" step after a workflow has been submitted.  

## Diagnostics

What Kubernetes provider are you using? 
EKS 1.18

What version of Argo Workflows are you running? 
v2.12.10

```yaml
Paste a workflow that reproduces the bug, including status:
kubectl get wf -o yaml ${workflow} 


# From my reproducible experiment 
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  creationTimestamp: "2021-03-29T20:55:26Z"
  generateName: steps-
  generation: 3
  labels:
    workflows.argoproj.io/phase: Running
  managedFields:
  - apiVersion: argoproj.io/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:generateName: {}
      f:spec:
        .: {}
        f:arguments:
          .: {}
          f:parameters: {}
        f:entrypoint: {}
        f:templates: {}
      f:status:
        .: {}
        f:finishedAt: {}
    manager: argo
    operation: Update
    time: "2021-03-29T20:55:26Z"
  - apiVersion: argoproj.io/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:labels:
          .: {}
          f:workflows.argoproj.io/phase: {}
      f:spec:
        f:affinity:
          .: {}
          f:podAffinity:
            .: {}
            f:preferredDuringSchedulingIgnoredDuringExecution: {}
        f:nodeSelector:
          .: {}
          f:spot: {}
        f:parallelism: {}
        f:securityContext:
          .: {}
          f:fsGroup: {}
        f:serviceAccountName: {}
        f:tolerations: {}
        f:ttlStrategy:
          .: {}
          f:secondsAfterCompletion: {}
          f:secondsAfterFailure: {}
          f:secondsAfterSuccess: {}
      f:status:
        f:nodes:
          .: {}
          f:steps-tt9wq:
            .: {}
            f:children: {}
            f:displayName: {}
            f:finishedAt: {}
            f:id: {}
            f:name: {}
            f:phase: {}
            f:progress: {}
            f:startedAt: {}
            f:templateName: {}
            f:templateScope: {}
            f:type: {}
          f:steps-tt9wq-196468504:
            .: {}
            f:boundaryID: {}
            f:children: {}
            f:displayName: {}
            f:finishedAt: {}
            f:id: {}
            f:name: {}
            f:phase: {}
            f:progress: {}
            f:startedAt: {}
            f:templateName: {}
            f:templateScope: {}
            f:type: {}
          f:steps-tt9wq-4247131649:
            .: {}
            f:boundaryID: {}
            f:displayName: {}
            f:finishedAt: {}
            f:hostNodeName: {}
            f:id: {}
            f:inputs:
              .: {}
              f:parameters: {}
            f:message: {}
            f:name: {}
            f:phase: {}
            f:progress: {}
            f:startedAt: {}
            f:templateName: {}
            f:templateScope: {}
            f:type: {}
        f:phase: {}
        f:progress: {}
        f:startedAt: {}
    manager: workflow-controller
    operation: Update
    time: "2021-03-29T20:55:36Z"
  name: steps-tt9wq
  namespace: sandbox
  resourceVersion: "86320807"
  selfLink: /apis/argoproj.io/v1alpha1/namespaces/sandbox/workflows/steps-tt9wq
  uid: <uid>
spec:
  affinity:
    podAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
      - podAffinityTerm:
          labelSelector:
            matchExpressions:
            - key: workflows.argoproj.io/completed
              operator: In
              values:
              - "false"
          topologyKey: kubernetes.io/hostname
        weight: 100
      - podAffinityTerm:
          labelSelector:
            matchExpressions:
            - key: workflows.argoproj.io/workflow
              operator: Exists
          topologyKey: kubernetes.io/hostname
        weight: 1
  arguments:
    parameters:
    - name: message
      value: hi
  entrypoint: hello-hello-hello
  nodeSelector:
    spot: "true"
  parallelism: 1
  securityContext:
    fsGroup: 123456789
  serviceAccountName: argo-sa
  templates:
  - arguments: {}
    inputs: {}
    metadata: {}
    name: hello-hello-hello
    outputs: {}
    steps:
    - - arguments:
          parameters:
          - name: message
            value: hello1
        name: hello1
        template: whalesay
    - - arguments:
          parameters:
          - name: message
            value: hello2a
        name: hello2a
        template: whalesay
      - arguments:
          parameters:
          - name: message
            value: hello2b
        name: hello2b
        template: whalesay
  - arguments: {}
    container:
      args:
      - '{{inputs.parameters.message}}'
      command:
      - cowsay
      image: docker/whalesay
      name: ""
      resources: {}
    inputs:
      parameters:
      - name: message
    metadata: {}
    name: whalesay
    outputs: {}
  tolerations:
  - effect: NoSchedule
    key: spot
    operator: Equal
    value: "true"
  ttlStrategy:
    secondsAfterCompletion: 3600
    secondsAfterFailure: 3600
    secondsAfterSuccess: 3600
status:
  finishedAt: null
  nodes:
    steps-tt9wq:
      children:
      - steps-tt9wq-196468504
      displayName: steps-tt9wq
      finishedAt: null
      id: steps-tt9wq
      name: steps-tt9wq
      phase: Running
      progress: 0/1
      startedAt: "2021-03-29T20:55:26Z"
      templateName: hello-hello-hello
      templateScope: local/steps-tt9wq
      type: Steps
    steps-tt9wq-196468504:
      boundaryID: steps-tt9wq
      children:
      - steps-tt9wq-4247131649
      displayName: '[0]'
      finishedAt: null
      id: steps-tt9wq-196468504
      name: steps-tt9wq[0]
      phase: Running
      progress: 0/1
      startedAt: "2021-03-29T20:55:26Z"
      templateName: hello-hello-hello
      templateScope: local/steps-tt9wq
      type: StepGroup
    steps-tt9wq-4247131649:
      boundaryID: steps-tt9wq
      displayName: hello1
      finishedAt: null
      hostNodeName: <redacted>
      id: steps-tt9wq-4247131649
      inputs:
        parameters:
        - name: message
          value: hello1
      message: ContainerCreating
      name: steps-tt9wq[0].hello1
      phase: Pending
      progress: 0/1
      startedAt: "2021-03-29T20:55:26Z"
      templateName: whalesay
      templateScope: local/steps-tt9wq
      type: Pod
  phase: Running
  progress: 0/1
  startedAt: "2021-03-29T20:55:26Z"

```

```
Paste the logs from the workflow controller:
kubectl logs -n argo $(kubectl get pods -l app=workflow-controller -n argo -o name) | grep ${workflow}

NOTE:  Max parallelism is set to 100.  I successfully ran this workflow pre- and post- capture of the logs without any other workflows being submitted. 

 time="2021-03-29T20:55:26.792Z" level=info msg="Processing workflow" namespace=sandbox workflow=steps-tt9wq
 time="2021-03-29T20:55:26.793Z" level=info msg="Updated phase  -> Running" namespace=sandbox workflow=steps-tt9wq
 time="2021-03-29T20:55:26.794Z" level=info msg="Steps node steps-tt9wq initialized Running" namespace=sandbox workflow=steps-tt9wq
 time="2021-03-29T20:55:26.794Z" level=info msg="StepGroup node steps-tt9wq-196468504 initialized Running" namespace=sandbox workflow=steps-tt9wq
 time="2021-03-29T20:55:26.794Z" level=info msg="Pod node steps-tt9wq-4247131649 initialized Pending" namespace=sandbox workflow=steps-tt9wq
 time="2021-03-29T20:55:26.824Z" level=info msg="Created pod: steps-tt9wq[0].hello1 (steps-tt9wq-4247131649)" namespace=sandbox workflow=steps-tt9wq
 time="2021-03-29T20:55:26.824Z" level=info msg="Workflow step group node steps-tt9wq-196468504 not yet completed" namespace=sandbox workflow=steps-tt9wq
 time="2021-03-29T20:55:26.838Z" level=info msg="Workflow update successful" namespace=sandbox phase=Running resourceVersion=86320701 workflow=steps-tt9wq
 time="2021-03-29T20:55:36.792Z" level=info msg="Processing workflow" namespace=sandbox workflow=steps-tt9wq
 time="2021-03-29T20:55:36.793Z" level=info msg="Updating node steps-tt9wq-4247131649 message: ContainerCreating"
 time="2021-03-29T20:55:36.794Z" level=info msg="workflow active pod spec parallelism reached 1/1" namespace=sandbox workflow=steps-tt9wq
 time="2021-03-29T20:55:36.794Z" level=error msg="error in entry template execution" error="Max parallelism reached" namespace=sandbox workflow=steps-tt9wq
 time="2021-03-29T20:55:36.809Z" level=info msg="Workflow update successful" namespace=sandbox phase=Running resourceVersion=86320807 workflow=steps-tt9wq
 time="2021-03-29T20:55:46.811Z" level=info msg="Processing workflow" namespace=sandbox workflow=steps-tt9wq
 time="2021-03-29T20:55:46.812Z" level=info msg="Workflow pod is missing" namespace=sandbox nodeName="steps-tt9wq[0].hello1" nodePhase=Pending recentlyStarted=false workflow=steps-tt9wq
 time="2021-03-29T20:55:46.813Z" level=info msg="workflow active pod spec parallelism reached 1/1" namespace=sandbox workflow=steps-tt9wq
 time="2021-03-29T20:55:46.813Z" level=error msg="error in entry template execution" error="Max parallelism reached" namespace=sandbox workflow=steps-tt9wq
 time="2021-03-29T21:15:46.812Z" level=info msg="Processing workflow" namespace=sandbox workflow=steps-tt9wq
 time="2021-03-29T21:15:46.813Z" level=info msg="Workflow pod is missing" namespace=sandbox nodeName="steps-tt9wq[0].hello1" nodePhase=Pending recentlyStarted=false workflow=steps-tt9wq
 time="2021-03-29T21:15:46.813Z" level=info msg="workflow active pod spec parallelism reached 1/1" namespace=sandbox workflow=steps-tt9wq
 time="2021-03-29T21:15:46.813Z" level=error msg="error in entry template execution" error="Max parallelism reached" namespace=sandbox workflow=steps-tt9wq
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by alexec** [comment]
Maybe this is some interplay between parallelism and created nodes. However, I'm surprised we have not had a any other complaints about this as I would imagine it would pretty common.

**Comment by alexec** [comment]
@rysunder can you please run the workflow without parallelism (or until running nodes complete) and see if the third pod runs?

**Comment by rysunder** [comment]
> @rysunder can you please run the workflow without parallelism (or until running nodes complete) and see if the third pod runs?

@alexec We originally ran it without Max Parallelism set.  We tried to fix the issue by setting Max Parallelism to 100.  If that is what you meant by "without parallelism" then we have already tested that with the same effect.  Also "parallelism reached 1/1" is mentioned multiple times in the logs.  

Is there any other information that may be helpful?

**Comment by alexec** [comment]
the controller will not schedule any pods where you reach parallelism - the example you attached has reached max parallelism - I want to know if disabling it allows those pods to scheduled which will help shed light on whether or not it is a bug with that feature, or something else

**Comment by alexec** [comment]
@sarabala1979 there may be two issues here:

* Is max parallelism preventing scheduling until running pods complete? It's not clear to me if there is actually an issue here.
* Is there some issue with ignoring or missing `ContainerCreating`? Those pods will be in `Pending` state.

**Comment by rysunder** [comment]
> the controller will not schedule any pods where you reach parallelism - the example you attached has reached max parallelism - I want to know if disabling it allows those pods to scheduled which will help shed light on whether or not it is a bug with that feature, or something else

@alexec Thank you for clarifying!  

I removed parallelism from the `controller.parallelism` and `controller.workflowDefaults.spec.parallelism` on the EKS cluster and re-ran the tests.

Interestingly, the workflow did not hang indefinitely.  It would pause when the pod was yanked, but eventually recovered (usually within a handful of seconds) and complete the workflow successfully.  Below are the workflow-controller logs related to the workflow.

```
Paste the logs from the workflow controller:
kubectl logs -n argo $(kubectl get pods -l app=workflow-controller -n argo -o name) | grep ${workflow}

 time="2021-03-31T21:31:37.357Z" level=info msg="Processing workflow" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:31:37.357Z" level=info msg="Updated phase  -> Running" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:31:37.359Z" level=info msg="Steps node steps-98j8q initialized Running" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:31:37.359Z" level=info msg="StepGroup node steps-98j8q-1810041543 initialized Running" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:31:37.360Z" level=info msg="Pod node steps-98j8q-1363360536 initialized Pending" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:31:37.388Z" level=info msg="Created pod: steps-98j8q[0].hello1 (steps-98j8q-1363360536)" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:31:37.388Z" level=info msg="Workflow step group node steps-98j8q-1810041543 not yet completed" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:31:37.402Z" level=info msg="Workflow update successful" namespace=sandbox phase=Running resourceVersion=87577095 workflow=steps-98j8q
 time="2021-03-31T21:31:47.357Z" level=info msg="Processing workflow" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:31:47.358Z" level=info msg="Workflow pod is missing" namespace=sandbox nodeName="steps-98j8q[0].hello1" nodePhase=Pending recentlyStarted=false workflow=steps-98j8q
 time="2021-03-31T21:31:47.384Z" level=info msg="Created pod: steps-98j8q[0].hello1 (steps-98j8q-1363360536)" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:31:47.384Z" level=info msg="Workflow step group node steps-98j8q-1810041543 not yet completed" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:31:57.386Z" level=info msg="Processing workflow" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:31:57.387Z" level=info msg="Setting node steps-98j8q-1363360536 outputs"
 time="2021-03-31T21:31:57.387Z" level=info msg="Updating node steps-98j8q-1363360536 status Pending -> Running"
 time="2021-03-31T21:31:57.388Z" level=info msg="Workflow step group node steps-98j8q-1810041543 not yet completed" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:31:57.404Z" level=info msg="Workflow update successful" namespace=sandbox phase=Running resourceVersion=87577245 workflow=steps-98j8q
 time="2021-03-31T21:32:07.406Z" level=info msg="Processing workflow" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:07.407Z" level=info msg="Updating node steps-98j8q-1363360536 status Running -> Succeeded"
 time="2021-03-31T21:32:07.408Z" level=info msg="Step group node steps-98j8q-1810041543 successful" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:07.408Z" level=info msg="node steps-98j8q-1810041543 phase Running -> Succeeded" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:07.408Z" level=info msg="node steps-98j8q-1810041543 finished: 2021-03-31 21:32:07.408360433 +0000 UTC" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:07.408Z" level=info msg="StepGroup node steps-98j8q-1877299114 initialized Running" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:07.408Z" level=info msg="SG Outbound nodes of steps-98j8q-1363360536 are [steps-98j8q-1363360536]" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:07.408Z" level=info msg="Pod node steps-98j8q-841259073 initialized Pending" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:07.431Z" level=info msg="Created pod: steps-98j8q[1].hello2a (steps-98j8q-841259073)" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:07.433Z" level=info msg="Pod node steps-98j8q-790926216 initialized Pending" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:07.457Z" level=info msg="Created pod: steps-98j8q[1].hello2b (steps-98j8q-790926216)" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:07.457Z" level=info msg="Workflow step group node steps-98j8q-1877299114 not yet completed" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:07.474Z" level=info msg="Workflow update successful" namespace=sandbox phase=Running resourceVersion=87577320 workflow=steps-98j8q
 time="2021-03-31T21:32:07.500Z" level=info msg="Labeled pod completed" namespace=sandbox pod=steps-98j8q-1363360536
 time="2021-03-31T21:32:17.432Z" level=info msg="Processing workflow" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:17.434Z" level=info msg="Setting node steps-98j8q-841259073 outputs"
 time="2021-03-31T21:32:17.434Z" level=info msg="Updating node steps-98j8q-841259073 status Pending -> Succeeded"
 time="2021-03-31T21:32:17.434Z" level=info msg="Workflow pod is missing" namespace=sandbox nodeName="steps-98j8q[1].hello2b" nodePhase=Pending recentlyStarted=false workflow=steps-98j8q
 time="2021-03-31T21:32:17.435Z" level=info msg="SG Outbound nodes of steps-98j8q-1363360536 are [steps-98j8q-1363360536]" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:17.467Z" level=info msg="Created pod: steps-98j8q[1].hello2b (steps-98j8q-790926216)" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:17.467Z" level=info msg="Workflow step group node steps-98j8q-1877299114 not yet completed" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:17.482Z" level=info msg="Workflow update successful" namespace=sandbox phase=Running resourceVersion=87577412 workflow=steps-98j8q
 time="2021-03-31T21:32:17.511Z" level=info msg="Labeled pod completed" namespace=sandbox pod=steps-98j8q-841259073
 time="2021-03-31T21:32:27.470Z" level=info msg="Processing workflow" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:27.472Z" level=info msg="Setting node steps-98j8q-790926216 outputs"
 time="2021-03-31T21:32:27.472Z" level=info msg="Updating node steps-98j8q-790926216 status Pending -> Succeeded"
 time="2021-03-31T21:32:27.473Z" level=info msg="SG Outbound nodes of steps-98j8q-1363360536 are [steps-98j8q-1363360536]" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:27.473Z" level=info msg="Step group node steps-98j8q-1877299114 successful" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:27.473Z" level=info msg="node steps-98j8q-1877299114 phase Running -> Succeeded" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:27.473Z" level=info msg="node steps-98j8q-1877299114 finished: 2021-03-31 21:32:27.473766129 +0000 UTC" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:27.473Z" level=info msg="Outbound nodes of steps-98j8q-841259073 is [steps-98j8q-841259073]" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:27.473Z" level=info msg="Outbound nodes of steps-98j8q-790926216 is [steps-98j8q-790926216]" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:27.473Z" level=info msg="Outbound nodes of steps-98j8q is [steps-98j8q-841259073 steps-98j8q-790926216]" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:27.473Z" level=info msg="node steps-98j8q phase Running -> Succeeded" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:27.473Z" level=info msg="node steps-98j8q finished: 2021-03-31 21:32:27.473908032 +0000 UTC" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:27.473Z" level=info msg="Checking daemoned children of steps-98j8q" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:27.473Z" level=info msg="Updated phase Running -> Succeeded" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:27.473Z" level=info msg="Marking workflow completed" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:27.473Z" level=info msg="Marking workflow as pending archiving" namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:27.474Z" level=info msg="Checking daemoned children of " namespace=sandbox workflow=steps-98j8q
 time="2021-03-31T21:32:27.493Z" level=info msg="Workflow update successful" namespace=sandbox phase=Succeeded resourceVersion=87577494 workflow=steps-98j8q
 time="2021-03-31T21:32:27.498Z" level=info msg="archiving workflow" namespace=sandbox uid=<uid> workflow=steps-98j8q
 time="2021-03-31T21:32:27.512Z" level=info msg="Labeled pod completed" namespace=sandbox pod=steps-98j8q-790926216
 time="2021-03-31T21:32:27.550Z" level=info msg="archiving workflow" namespace=sandbox uid=<uid> workflow=steps-98j8q
 time="2021-03-31T21:32:27.555Z" level=info msg="Queueing Succeeded workflow sandbox/steps-98j8q for delete in 1h0m0s"
 time="2021-03-31T21:32:27.556Z" level=info msg="Queueing Succeeded workflow sandbox/steps-98j8q for delete in 1h0m0s"
```

The next test I ran was to disable just the `controller.workflowDefaults.spec.parallelism` and leave the `controller.parallelism` set to what it was originally.  This resulted in hanging jobs!  

```
Paste the logs from the workflow controller:
kubectl logs -n argo $(kubectl get pods -l app=workflow-controller -n argo -o name) | grep ${workflow}

 time="2021-04-01T17:34:31.828Z" level=info msg="Processing workflow" namespace=sandbox workflow=steps-hrn66
 time="2021-04-01T17:34:31.829Z" level=info msg="Updated phase  -> Running" namespace=sandbox workflow=steps-hrn66
 time="2021-04-01T17:34:31.831Z" level=info msg="Steps node steps-hrn66 initialized Running" namespace=sandbox workflow=steps-hrn66
 time="2021-04-01T17:34:31.831Z" level=info msg="StepGroup node steps-hrn66-352772121 initialized Running" namespace=sandbox workflow=steps-hrn66
 time="2021-04-01T17:34:31.831Z" level=info msg="Pod node steps-hrn66-3293355422 initialized Pending" namespace=sandbox workflow=steps-hrn66
 time="2021-04-01T17:34:31.866Z" level=info msg="Created pod: steps-hrn66[0].hello1 (steps-hrn66-3293355422)" namespace=sandbox workflow=steps-hrn66
 time="2021-04-01T17:34:31.866Z" level=info msg="Workflow step group node steps-hrn66-352772121 not yet completed" namespace=sandbox workflow=steps-hrn66
 time="2021-04-01T17:34:31.879Z" level=info msg="Workflow update successful" namespace=sandbox phase=Running resourceVersion=88816984 workflow=steps-hrn66
 time="2021-04-01T17:34:41.802Z" level=info msg="Processing workflow" namespace=sandbox workflow=steps-hrn66
 time="2021-04-01T17:34:41.804Z" level=info msg="Updating node steps-hrn66-3293355422 message: ContainerCreating"
 time="2021-04-01T17:34:41.804Z" level=info msg="workflow active pod spec parallelism reached 1/1" namespace=sandbox workflow=steps-hrn66
 time="2021-04-01T17:34:41.804Z" level=error msg="error in entry template execution" error="Max parallelism reached" namespace=sandbox workflow=steps-hrn66
 time="2021-04-01T17:34:41.815Z" level=info msg="Workflow update successful" namespace=sandbox phase=Running resourceVersion=88817082 workflow=steps-hrn66
 time="2021-04-01T17:34:51.817Z" level=info msg="Processing workflow" namespace=sandbox workflow=steps-hrn66
 time="2021-04-01T17:34:51.819Z" level=info msg="Workflow pod is missing" namespace=sandbox nodeName="steps-hrn66[0].hello1" nodePhase=Pending recentlyStarted=false workflow=steps-hrn66
 time="2021-04-01T17:34:51.819Z" level=info msg="workflow active pod spec parallelism reached 1/1" namespace=sandbox workflow=steps-hrn66
 time="2021-04-01T17:34:51.819Z" level=error msg="error in entry template execution" error="Max parallelism reached" namespace=sandbox workflow=steps-hrn66
```

Finally, I disabled the `controller.parallelism` and re-enabled the `controller.workflowDefaults.spec.parallelism`.  The workflow recovered successfully when the pod or node was removed during the ContainerCreating step.  

In summary, it appears to be an issue related to the `controller.workflowsDefaults.spec.parallelism` that results in zombie workflows when a pod/node is removed during the ContainerCreating step.

**Comment by alexec** [comment]
Can you please share your workflows-controller-configmap?

**Comment by rysunder** [comment]
> Can you please share your workflows-controller-configmap?

Can do!

```yaml
 Name:         argo-workflow-controller-configmap
 Namespace:    argo
 Labels:       app.kubernetes.io/managed-by=Helm
               chart=argo-0.15.4
               heritage=Helm
               release=argo
 Annotations:  meta.helm.sh/release-name: argo
               meta.helm.sh/release-namespace: argo

 Data
 ====
 config:
 ----
 containerRuntimeExecutor: pns
 parallelism: 100
 artifactRepository:
   archiveLogs: true
   s3:
     bucket: <my-redacted-bucket>
     endpoint: s3.amazonaws.com
     insecure: false
     keyFormat: dev/{{workflow.creationTimestamp.Y}}/{{workflow.creationTimestamp.m}}/{{workflow.creationTimestamp.d}}/{{workflow.name}}/{{pod.name}}
     region: <region>
     useSDKCreds: true
 metricsConfig:
   enabled: true
   metricsTTL: 240m
   path: /metrics
   port: 8080
 persistence:
   archive: true
   archiveTTL: 90d
   connectionPool:
     maxIdleConns: 100
     maxOpenConns: 0
   mysql:
     database: argo_database
     host: <database-server-hostname>
     passwordSecret:
       key: password
       name: argo-mysql-config
     port: 3306
     tableName: argo_workflows
     userNameSecret:
       key: username
       name: argo-mysql-config
   nodeStatusOffLoad: false
 workflowDefaults:
   spec:
     affinity:
       podAffinity:
         preferredDuringSchedulingIgnoredDuringExecution:
         - podAffinityTerm:
             labelSelector:
               matchExpressions:
               - key: workflows.argoproj.io/completed
                 operator: In
                 values:
                 - "false"
             topologyKey: kubernetes.io/hostname
           weight: 100
         - podAffinityTerm:
             labelSelector:
               matchExpressions:
               - key: workflows.argoproj.io/workflow
                 operator: Exists
             topologyKey: kubernetes.io/hostname
           weight: 1
     nodeSelector:
       spot: "true"
     parallelism: 1
     securityContext:
       fsGroup: 123456789
     serviceAccountName: argo-sa
     tolerations:
     - effect: NoSchedule
       key: value
       operator: Equal
       value: "true"
     ttlStrategy:
       secondsAfterCompletion: 3600
       secondsAfterFailure: 3600
       secondsAfterSuccess: 3600
 sso:
   clientId:
     key: client-id
     name: argo-server-sso
   clientSecret:
     key: client-secret
     name: argo-server-sso
   issuer: https://login.microsoftonline.com/uid/v2.0
   rbac:
     enabled: true
   redirectUrl: https://argo-hostname.net/oauth2/callback
   scopes:
   - uid2/.default
   - email

 Events:  <none>
```

**Comment by alexec** [comment]
I'm not clear, there are two parallelism settings, `data.parallelism` and `data.workflowDefaults.spec.parallelism'. Removing which fixed the problem?

**Comment by rysunder** [comment]
> I'm not clear, there are two parallelism settings, `data.parallelism` and `data.workflowDefaults.spec.parallelism'. Removing which fixed the problem?

Removing `data.workflowDefaults.spec.parallelism` fixed the issue

**Comment by alexec** [comment]
Ok. Can you attach:

1. `status.storedWorkflowTemplateSpec` for the bad workflow.
2. `status.storedWorkflowTemplateSpec` for the good workflow.

**Comment by rysunder** [comment]
> Ok. Can you attach:
> 
> 1. `status.storedWorkflowTemplateSpec` for the bad workflow.
> 2. `status.storedWorkflowTemplateSpec` for the good workflow.

More than happy to ... if I knew how to get that information.  Is there a quick how-to on it I could follow?

**Comment by sarabala1979** [comment]
@alexec problem is the active pod is calculated based on node status (pending and running) In this scenario, the node  status is in pending status and parallelism is 1, so the controller is the exit from the parallel check.  `podReconciliation` is just continue if pod is missing and node status is pending.

Suggestions: 
1. Introduce a new state for pod missing which needs to exclude from active pod count.

**Comment by sarabala1979** [comment]
https://github.com/argoproj/argo-workflows/blob/4e450e250168e6b4d51a126b784e90b11a0162bc/workflow/controller/operator.go#L969-L980

https://github.com/argoproj/argo-workflows/blob/4e450e250168e6b4d51a126b784e90b11a0162bc/workflow/controller/node_counters.go#L9-L18
